notes/patching-system.md

```diff
@@ -0,0 +1,100 @@
+# The Patching System
+
+This document explains how Shrink Ray efficiently applies multiple changes in parallel.
+
+## The Problem
+
+When reducing, we often want to try many potential changes:
+- Delete each of 1000 lines
+- Delete each token
+- Replace each integer with a smaller value
+
+Trying these one-at-a-time is slow. But we can often apply multiple changes simultaneously if they don't conflict.
+
+## Patches Abstract Interface
+
+The `Patches[Patch, Target]` class defines how patches work:
+
+```python
+class Patches[Patch, Target]:
+    @property
+    def empty(self) -> Patch: ...           # The identity patch
+    def combine(self, *patches) -> Patch: ...  # Merge patches
+    def apply(self, patch, target) -> Target: ...  # Apply patch to target
+    def size(self, patch) -> int: ...       # How much the patch reduces
+```
+
+### Cuts (Most Common)
+
+`Cuts` patches are lists of `(start, end)` intervals to delete:
+```python
+patch = [(0, 5), (10, 15)]  # Delete bytes 0-5 and 10-15
+```
+
+Combining cuts merges their intervals. The `size` is total bytes deleted.
+
+### Other Patch Types
```

Comment:
There are quite a few other Patch types. `grep -R '^class.\+(Patches' src` will give you a complete list.